pow/src/main/java/tech/pegasys/artemis/pow/DepositRequestManager.java

```diff
@@ -0,0 +1,222 @@
+        err);
+
+    asyncRunner
+        .getDelayedFuture(Constants.ETH1_DEPOSIT_REQUEST_RETRY_TIMEOUT, TimeUnit.SECONDS)
+        .finish(
+            this::getLatestDeposits,
+            (error) -> LOG.warn("Unable to execute delayed request. Dropping request", error));
+  }
+
+  private SafeFuture<Void> fetchAndPublishDepositEventsInBlockRange(
+      DefaultBlockParameter fromBlock, DefaultBlockParameter toBlock) {
+    SafeFuture<List<DepositContract.DepositEventEventResponse>> eventsFuture =
+        depositContract.depositEventEventsInRange(fromBlock, toBlock);
+
+    return eventsFuture
+        .thenApply(this::groupDepositEventResponsesByBlockHash)
+        .thenCompose(
+            groupedDepositEventResponsesByBlockHash -> {
+              Map<String, SafeFuture<EthBlock.Block>> neededBlocksByHash =
+                  getMapOfEthBlockFutures(groupedDepositEventResponsesByBlockHash.keySet());
+              return SafeFuture.allOf(neededBlocksByHash.values().toArray(SafeFuture[]::new))
+                  .thenApply(
+                      done ->
+                          constructDepositsFromBlockEvents(
+                              neededBlocksByHash, groupedDepositEventResponsesByBlockHash));
+            })
+        .thenAccept(
+            (depositsFromBlockEventList) ->
+                depositsFromBlockEventList.stream()
+                    .sorted(Comparator.comparing(DepositsFromBlockEvent::getBlockNumber))
+                    .forEachOrdered(this::publishDeposits));
+  }
+
+  private List<DepositsFromBlockEvent> constructDepositsFromBlockEvents(
+      Map<String, SafeFuture<EthBlock.Block>> blockFutureByBlockHash,
+      Map<String, List<DepositContract.DepositEventEventResponse>> groupedDepositEventResponses) {
+    return groupedDepositEventResponses.entrySet().stream()
+        .map(
+            (entry) -> {
+              String blockHash = entry.getKey();
+              List<DepositContract.DepositEventEventResponse> groupedDepositEventResponse =
+                  entry.getValue();
+              EthBlock.Block block =
+                  checkNotNull(blockFutureByBlockHash.get(blockHash).getNow(null));
+
+              return new DepositsFromBlockEvent(
+                  UnsignedLong.valueOf(block.getNumber()),
+                  Bytes32.fromHexString(block.getHash()),
+                  UnsignedLong.valueOf(block.getTimestamp()),
+                  groupedDepositEventResponse.stream()
+                      .map(Deposit::new)
+                      .sorted(Comparator.comparing(Deposit::getMerkle_tree_index))
+                      .collect(Collectors.toList()));
+            })
+        .collect(Collectors.toList());
+  }
+
+  private Map<String, SafeFuture<EthBlock.Block>> getMapOfEthBlockFutures(
+      Set<String> neededBlockHashes) {
+    return neededBlockHashes.stream()
+        .collect(
+            Collectors.toUnmodifiableMap(blockHash -> blockHash, eth1Provider::getEth1BlockFuture));
```

Comment:
really small nit: I'd probably use static imports for all the methods in `Collectors` just for brevity.

Reply:
done.